### ISSUE TITLE
Loading a chat the first time with files (or images sent as files without compression) causes a long UI freeze

### ISSUE THREAD

**Issue by Gira-X** [zevlg/telega.el#281]
I tried disabling the `telega-webpage-*` options via `M-x customize-variable` like this:

![Screen Shot 2021-04-03 at 23 12 41](https://user-images.githubusercontent.com/15002298/113491591-4a912000-94d2-11eb-82ff-0592042c5168.png)

But in Telega, there is still a preview:

![image](https://user-images.githubusercontent.com/15002298/113491609-6dbbcf80-94d2-11eb-86f5-53fee4ca7177.png)

---

I have the feeling that scraping the URLs delays loading the chat (especially chats with lots of links) initially for multiple seconds, so I would love to disable this feature altogether.

Every time I send a link, it has to generate the URL preview and completely blocks the UI apparently. One can test this by sending 2 messages, one with just some text like `oeu` which is instant; and the second one with an URL link which exhibits a noticeable delay.

I am on `telega v0.7.21 (TDLib v1.7.2) (telega-server v0.7.6)`.

**Comment by zevlg** [comment]
If you feel telega does some heavy things, please set `debug-on-quit` to non-nil and press `C-g` on heavy telega operations to get the backtrace.

You might be hitting https://lists.gnu.org/archive/html/bug-gnu-emacs/2020-01/msg00548.html Emacs bug as well

Anyway, if you want to disable web page preview, set `telega-webpage-preview-description-limit` to 0 as described in docstring

**Comment by zevlg** [comment]
Also, what is your Emacs version, to reproduce the behvoiur?

Does this reproduces on `emacs -q` run?

**Comment by Gira-X** [comment]
I am on `GNU Emacs 27.2 (build 1, x86_64-apple-darwin18.5.0, Carbon Version 158 AppKit 1671.4) of 2021-04-02` but I remember it being present on earlier 27.X versions as well.

> Does this reproduces on emacs -q run?

I broke the `emacs` CLI command by linking the incorrect program and haven't been able to fix it correctly yet, so I can't test this right now :(

> Anyway, if you want to disable web page preview, set telega-webpage-preview-description-limit to 0 as described in docstring

I have it set to 0 and it does not work. See my screenshot above, it still displays a webpage preview. Is there a way to completely disable it?

---

With `debug-on-quit`, I was not able to get any meaningful backtrace, so I tried the `profiler` and got this:

This is from opening a single chat which has a lot of links:

```
- server-process-filter                                           256  52%
 - server-execute-continuation                                    256  52%
  - #<compiled 0x1ff5b046b5c9>                                    256  52%
   - server-execute                                               255  51%
    - server-delete-client                                        184  37%
     - delete-process                                             184  37%
      - redisplay_internal (C function)                            11   2%
       - find-image                                                11   2%
        - image-search-load-path                                   11   2%
           image-search-load-path                                  11   2%
    - mapc                                                         71  14%
     - funcall                                                     71  14%
      + #<compiled 0x1ff5b046b58d>                                 71  14%
+ redisplay_internal (C function)                                 100  20%
- telega-server--filter                                            66  13%
 - telega-server--parse-commands                                   66  13%
  - apply                                                          66  13%
   - telega-server--dispatch-cmd                                   66  13%
    - #<compiled 0x1ff5b02b622d>                                   66  13%
     - telega-chatbuf--insert-messages                             50  10%
      - seq-do                                                     50  10%
       - mapc                                                      50  10%
        - #<compiled 0x1ff5b0473231>                               50  10%
         - ewoc-enter-after                                        49   9%
          - ewoc-enter-before                                      49   9%
           - ewoc--insert-new-node                                 49   9%
            - ewoc--refresh-node                                   49   9%
             - #<compiled 0x1ff5b08b095d>                          49   9%
              - telega-msg--pp                                     49   9%
               - telega-button--insert                             47   9%
                - telega-ins--message                              35   7%
                 - apply                                           35   7%
                  - telega-ins--message0                           35   7%
                   + telega-ins--content                           16   3%
                   + telega-fmt-eval-attrs                          7   1%
                   + telega-user-avatar-image                       4   0%
                   + telega-ins--message-header                     4   0%
                   + telega-ins--outgoing-status                    1   0%
                     generate-new-buffer                            1   0%
                   + telega-ins--fwd-info-inline                    1   0%
                   + telega-ins--image                              1   0%
                + telega-ins--message-no-header                    11   2%
               + telega-ins--message-header                         1   0%
               + telega-chat-match-p                                1   0%
         - telega-fmt-text                                          1   0%
          - telega-string-fmt-text-length                           1   0%
             count-matches                                          1   0%
     + #<compiled 0x1ff5b097728d>                                  16   3%
- ...                                                              61  12%
   Automatic GC                                                    61  12%
+ timer-event-handler                                               6   1%
+ command-execute                                                   3   0%
```

This is the state on opening a chat for the first time and it stays there for a few seconds when there are a lot of chat links for me:

In this state the UI is entirely frozen.

![image](https://user-images.githubusercontent.com/15002298/113499289-125f0100-9515-11eb-8132-85653361795e.png)

**Comment by Gira-X** [comment]
Using the code below, the webpage preview is successfully removed which is great, but I see absolutely no performance improvements as I hoped :(

```elisp
(defun telega-ins--webpage-own (msg &optional web-page) nil)
(advice-add 'telega-ins--webpage :override #'telega-ins--webpage-own)
```

It looks like this now:

![image](https://user-images.githubusercontent.com/15002298/113502038-df733800-9529-11eb-9141-74d1d765645c.png)

---

Any clue why chats with lots of links still freeze Emacs the first time they are opened?

**Comment by zevlg** [comment]
> Any clue why chats with lots of links still freeze Emacs the first time they are opened?

Check you don’t hit into this Emacs bug - https://lists.gnu.org/archive/html/bug-gnu-emacs/2020-01/msg00548.html

**Comment by Gira-X** [comment]
> Check you don’t hit into this Emacs bug - https://lists.gnu.org/archive/html/bug-gnu-emacs/2020-01/msg00548.html

I evaluated `(insert "a\xfe0f")` in the scratch buffer and it has absolutely no impact on scrolling speed.

---

Once the chat is fully loaded, scrolling through the chats is perfectly fine performance-wise.

**Comment by zevlg** [comment]
Dunno, what is the reason, but looks similar to fe0f bug

Try to set `telega-chat-history-limit` to less value, to load/format less messages

Does anybody else has same behaviour?

**Comment by Gira-X** [comment]
Thanks, this actually helped (somewhat)!

I had it set to `12` already, so I lowered it even more to `6` and surprisingly the chat loads very fast now on initial launch which is great because I can start typing without having to wait for the `history loading` to finish freezing up the UI.

It's just that on scrolling up it still gets the same lock-up once it keeps on loading old chat messages.
I suspect there is some message content just in this one chat of mine causing the lag but I can't pinpoint it yet.

I'll keep you updated should I find the root cause!

**Comment by Gira-X** [comment]
I figured it out!

The long UI freeze happens (at least for me) when a chat has image messages which are sent without image compression. The same freeze happens on other file types (like pdf files).

Even with a 400KB file the UI is frozen for seconds.

Is this a bug I found? Or is this just how telega behaves for such files?

**Comment by Gira-X** [comment]
The UI freeze does not seem to be dependent on the file size as well. I just tested with a 500 byte file which froze the UI as long as the 400KB file!

**Comment by zevlg** [comment]
So, your Emacs is slow at displaying images?

**Comment by Gira-X** [comment]
> So, your Emacs is slow at displaying images?

It is not. Images sent with compression display instantly on entering a chat and cause no scrolling problems.

The UI freeze only happens when files (or images sent without compression) are sent regardless of file size. It makes no difference if the file has an image preview.

---

So to reproduce:

- send a file like `test.org` (say 100 bytes with any content) to someone
- restart Emacs
- open the same chat where `test.org` was sent

For me the UI completely freezes for some seconds (and happens in every chat).
Does it not freeze for you?

---

I just now updated to `telega v0.7.35 (TDLib v1.7.3) (telega-server v0.7.7)` and it has the same problem.

**Comment by zevlg** [comment]
> For me the UI completely freezes for some seconds (and happens in every chat).
> Does it not freeze for you?
No, I can't reproduce this :(

Don't know what causes this behaviour or your side, not sure telega could do anything about this.  You are hitting some Emacs bug similar to  https://lists.gnu.org/archive/html/bug-gnu-emacs/2020-01/msg00548.html 

Try to shrink it to some `(insert xx)` commands that can be used in Emacs bug report

**Comment by zevlg** [comment]
Also make sure these freezes reproduces in `emacs -q` environment

**Comment by Gira-X** [comment]
While messing around, I found out that setting `M-x customize-option telega-symbol-attachment` to an empty string speeds up the history loading a significant amount! It still takes 1 or 2 seconds but that is acceptable compared to over 6 seconds before!

I also tested this in `brew install emacs-plus@28` and `brew install emacs-mac` (this one on 27.2) and they both exhibited this issue but at least I found a somewhat passable solution.

Might have been indeed been the Emacs bug you linked after all ;)
